File: 3rd_party_slots/rust_metta_bus_client/src/bus_node.rs
use tokio::runtime::Builder;
use tonic::{Request, Status};

use das_proto::{atom_space_node_client::AtomSpaceNodeClient, MessageData};

mod das_proto {
	tonic::include_proto!("dasproto");
}

use crate::{bus::Bus, types::BoxError};

#[derive(Debug, Default, Clone)]
pub struct BusNode {
	pub host_id: String,
	pub peer_id: String,
	pub bus: Bus,
	pub send_join_network: bool,
}

impl BusNode {
	pub fn new(
		host_id: String, commands: Vec<String>, target_id: String,
	) -> Result<Self, BoxError> {
		let mut bus = Bus::new();

		for command in commands {
			bus.add(command.clone());
		}

		Ok(Self { host_id, peer_id: target_id, bus, send_join_network: true })
	}

	pub fn node_id(&self) -> String {
		self.host_id.clone()
	}

	pub fn send_bus_command(&self, command: String, args: Vec<String>) -> Result<(), BoxError> {
		let target_id = self.bus.get_ownership(command.clone());
		if target_id.is_empty() {
			log::error!(target: "das", "Bus: no owner is defined for command <{command}>");
			Err("Bus: no owner is defined for command".into())
		} else {
			log::debug!(target: "das", "BUS node {} is routing command {command} to {target_id}", self.node_id());
			self.send(command, args, target_id.to_string(), false)
		}
	}

	pub fn send(
		&self, command: String, args: Vec<String>, target_id: String, is_broadcast: bool,
	) -> Result<(), BoxError> {
		let request = Request::new(MessageData {
			command,
			args,
			sender: self.host_id.clone(),
			is_broadcast,
			visited_recipients: vec![],
		});

		let runtime = Builder::new_multi_thread().enable_all().build().unwrap();
		runtime.block_on(async move {
			let target_addr = format!("http://{target_id}");
			log::trace!(target: "das", "BusNode::query(target_addr): {target_addr}");
			match AtomSpaceNodeClient::connect(target_addr).await {
				Ok(mut client) => client.execute_message(request).await,
				Err(err) => Err(Status::internal(format!("Client failed to connect: {err}"))),
			}
		})?;

		Ok(())
	}
}

File: 3rd_party_slots/python_client/hyperon_das/bus_node.py
import grpc

from hyperon_das._grpc.atom_space_node_pb2_grpc import AtomSpaceNodeStub
from hyperon_das._grpc import atom_space_node_pb2

from hyperon_das.bus import Bus
from hyperon_das.logger import log


class BusNode:
    def __init__(self, node_id: str, node_commands: list[str], known_peer: str = ""):
        self._bus = Bus()
        self.node_id = node_id
        for command in node_commands:
            self._bus.add(command)
            self._bus.set_ownership(command, known_peer)

    def send_bus_command(self, command: str, args: list[str]) -> None:
        target_id = self._bus.get_ownership(command)
        if target_id == "":
            raise RuntimeError(f"Bus: no owner is defined for command <{command}>")
        else:
            log.info(f"BUS node {self.node_id} is routing command <{command}> to {target_id}")
            self.send(command, args, target_id)

    def send(self, command: str, args: list[str], target_id: str) -> None:
        with grpc.insecure_channel(target_id) as channel:
            stub = AtomSpaceNodeStub(channel)
            message = atom_space_node_pb2.MessageData(
                command=command,
                args=args,
                sender=self.node_id,
                is_broadcast=False,
                visited_recipients=[]
            )
            try:
                stub.execute_message(message)
            except grpc.RpcError as e:
                log.error(f"Error: {e}")
